src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final float INCREASE_FACTOR = 1.5f;
+
+    private int size;
+    private int capacity = DEFAULT_CAPACITY;
+    private Object[] list;
+
+    public ArrayList() {
+        this.list = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndexForAdding(index);
+        if (size >= capacity) {
+            increaseCapacity();
+        }
+        if (index < size) {
+            shiftAllLeftFromIndex(index);
+        }
+        list[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        return (T) list[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndex(index);
+        list[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndex(index);
+        T removedEl = (T) list[index];
+        shiftAllRightFromIndex(index);
+        size--;
+        return removedEl;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedEl;
+        for (int i = 0; i < size; i++) {
+            if (element == null && list[i] == null
+                    || list[i] != null && list[i].equals(element)) {
+                removedEl = (T) list[i];
+                remove(i);
+                return removedEl;
+            }
+        }
+        throw new NoSuchElementException("No such element " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void increaseCapacity() {
+        Object[] tempList = list;
+
+        capacity *= INCREASE_FACTOR;
+        list = new Object[capacity];
+
+        for (int i = 0; i < size; i++) {
+            list[i] = tempList[i];
+        }
```

Comment:
It is preferable and simpler to utilize the System.arraycopy() method for array copying operations.